core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java

```diff
@@ -302,6 +302,29 @@ public static SqlCall stripSeparator(SqlCall call) {
   public static final SqlReturnTypeInference DATE_NULLABLE =
       DATE.andThen(SqlTypeTransforms.TO_NULLABLE);
 
+  /**
+   * Type-inference strategy that returns the type of the first operand,
+   * unless it is a DATE, in which case the return type is TIMESTAMP.
+   */
+  public static final SqlReturnTypeInference ARG0_EXCEPT_DATE = opBinding -> {
```

Comment:
Might be nice to add a comment referencing [CALCITE-5757] to remind that this is for BQ typing behavior

Reply:
Updated, let me know if it's clear enough